assets/hip-1037/Specification-Format-Style-Guidelines.md

```diff
@@ -0,0 +1,408 @@
+     because fields are output as markdown tables, and markdown tables require
+     each row to be a single long line.
+- Be _prescriptive_ in requirements.  Do not _describe_ the behavior.
+   - Declare what MUST be done for valid inputs, and what SHALL be present in
+     valid outputs (or in state messages).
+   - Declare what SHALL happen as a result of a successful, or failed,
+     transaction.
+   - The word `will` usually means `SHALL` or `MUST`.
+   - The word `is` often means `SHALL be`.
+   - Document both positive (`SHALL` or `MUST`) and negative (`SHALL NOT` or
+     `MUST NOT`) requirements.
+   - Avoid "perfect" or "imperfect" verb aspects and "wiggle" words
+     (might, could, perhaps, etc...).
+      - The specification items MAY, SHOULD, and RECOMMENDED serve to fill
+        the _occasional_ need for less imperative requirements.
+   - Also avoid "progressive" verbs (e.g. working, writing, going, deleting).
+- Requirements should be short and focused.
+   - Better to have three separate lines for three items than one longer
+     line with three clauses.
+   - Duplicating text across lines with slightly different requirements
+     is not only acceptable, but often recommended.
+   - Well written specification tends to be much more short and clipped
+     sentences than "normal" prose. Some describe the flow as "staccato".
+- Required and Optional fields
+   - The default state for any field is "optional", and all required fields
+     should be specified as `REQUIRED`. If the presence or absence of a field
+     must be clear _separate_ from the default value of that field, that field
+     should use proto3 "explicit presence" and be marked with
+     the `optional` keyword.
+- Add another blank line after requirements for `message` documents only.
+
+### Other elements
+- Each message that represents a transaction body **must** document the
+  Block Stream effects with a heading `### Block Stream Effects`.
+- General notes go last, for `message` only, with a heading
+  `#### Additional Notes`.
+- There are cases where description text for a message or field is
+  exceptionally detailed. In these cases it _may_ be appropriate to
+  separate broad categories of information with a single horizontal line
+  (i.e. an `<hr>` tag, `---` in markdown). These cases _should_ be quite
+  rare, however; we prefer to avoid the need for such large blocks of
+  specification text.
+- Line oriented comments (i.e. lines prefixed with `//`) may be used for
+  any comments needed in a proto file that are not intended for API
+  specification.  Examples include explaining an `oneof` block, or describing
+  why a particular field number or name is reserved.
+
+## Content Guidelines
+### General `proto` File Guidelines
+- Protocol buffers are compiled for the `hedera-services` codebase using a
+  custom processor 'PBJ'. While no other entity is required to use `PBJ`, all
```

Comment:
Hmm, is this essentially enforcing the community to have a comfortable understanding of PBJ outputs. This may be true but do we want to require this?
Should we call out the practical implications here so users know if they are compatible. How can they have confidence, does it require ensuring certain reserved types aren't used or does it require passing some tests on PBJ?

Reply:
The simplest way is to pass the PBJ integration test, or pass compilation in the `hedera-services` codebase.  Probably we need to, when resources permit, split the PBJ integration test out into a useable compatibility test library for general use.

The `hedera-services` repo is already the primary repository for any proto under the `services` subfolder (hedera-protobufs is a manually maintained copy), and the current direction is to continue to split every group of proto files out to the various code repositories (so Block Node would have block stream; mirror node would have MN their files, etc...).

Ideally, PBJ _should_ compile anything that `protoc` would.  There are a decent array of open bugs and limitations that prevent this, however.  PBJ is also behind protoc standard by a bit.  PBJ supports proto3 while protoc (with the new "editions" versioning) is at the `2023 Edition`.

Reply:
I've added some explanation to describe the recommended current process.